src/vs/workbench/browser/parts/editor/tabsTitleControl.ts

```diff
@@ -1437,13 +1438,14 @@ export class TabsTitleControl extends TitleControl {
 			let tabsWrapMultiLine = tabsContainer.classList.contains('wrap');
 
 			// Tabs do not wrap multiline: add wrapping if tabs exceed the tabs container width
-			if (!tabsWrapMultiLine && allTabsWidth > visibleTabsContainerWidth) {
+			// and the height of the tabs container does not exceed the maximum
+			if (!tabsWrapMultiLine && allTabsWidth > visibleTabsContainerWidth && (tabsContainer.offsetHeight <= TabsTitleControl.MAX_WRAPPED_HEIGHT)) {
 				tabsContainer.classList.add('wrap');
 				tabsWrapMultiLine = true;
 			}
 
-			// Tabs wrap multiline: remove wrapping if height exceeds available height
-			if (tabsWrapMultiLine && tabsContainer.offsetHeight > dimensions.available.height) {
+			// Tabs wrap multiline: remove wrapping if height exceeds the maximum
+			if (tabsWrapMultiLine && tabsContainer.offsetHeight > TabsTitleControl.MAX_WRAPPED_HEIGHT) {
```

Comment:
@SneakyFish5 shouldn't we check on both `dimensions.available.height` and `MAX_WRAPPED_HEIGHT`? What if the available height is less than `MAX_WRAPPED_HEIGHT`?

Reply:
Oh right good point, I re-added that in, thanks. Changed max amount of rows to 3 as well as requested.